File: leetcode/python/39_combination_sum.py
class Solution:
    def combinationSum(self, candidates: list[int], target: int) -> list[list[int]]:
        res: list[list[int]] = []
        candidates.sort()
        self.helper(candidates, target, 0, [], res)
        return res
    
    def helper(self, nums: list[int], target: int, i: int, comb: list[int], res: list[list[int]]) -> None:
        if target == 0:
            return res.append(comb.copy())
        
        for i in range(i, len(nums)):
            n = nums[i]
            if target - n < 0:
                return
            comb.append(n)
            self.helper(nums, target-n, i, comb, res)
            comb.pop()

if __name__ == '__main__':
    s = Solution()
    print(s.combinationSum([2,3,6,7], 7))
    print(s.combinationSum([2,3,5], 8))
    print(s.combinationSum([2], 1))

File: leetcode/c++/39_combination_sum.cpp
#include <vector>

using namespace std;

class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        sort(candidates.begin(), candidates.end());
        vector<vector<int>> res;
        helper(candidates, target, 0, {}, res);
        return res;
    }

    void helper(vector<int>& candidates, int target, int l, vector<int> comb, vector<vector<int>>& res) {
        if (target == 0) {
            return res.push_back(comb);
        }

        for (int i = l; i < candidates.size(); i++) {
            if (candidates[i] > target) {
                break;
            }
            comb.push_back(candidates[i]);
            helper(candidates, target - candidates[i], i, comb, res);
            comb.pop_back();
        }
    }
};

int main() {
    Solution s;
    vector<int> candidates = {2, 3, 6, 7};
    int target = 7;
    auto res = s.combinationSum(candidates, target);
    return 0;
}